asf_installer_survey/notebooks/section_6_analysis.py

```diff
@@ -0,0 +1,1670 @@
+#       jupytext_version: 1.11.2
+#   kernelspec:
+#     display_name: installersurvey
+#     language: python
+#     name: python3
+# ---
+
+# %% [markdown]
+# ## Section 6: Maintenance and servicing
+# **Aims:** Generating outputs for Section 6 (RQs 6.01-6.25) as planned in [Heat Pump Installer Survey data analysis plan](https://docs.google.com/document/d/1M1nzdf3fyTjipmaKJKViin0EB3e3R1afOQglwMGJaII/edit#heading=h.jzmlm8j0kyve)<br>
+# **Author:** Elysia Lucas<br>
+# **Date:** 2024-04-04<br>
+# <i>Update:</i> 2024-04-17 (Re-coding of free text responses)
+# <p> 
+# N.B. As of 11.04.2024, only CompanySizeOwner and CompanySizeEmployee outputs with version 2 categorisation are generated.
+
+# %% [markdown]
+# ### Importing and pre-processing
+
+# %%
+with open("data_setup.py") as file:
+    exec(file.read())
+
+with open("free_text_recode.py") as file:
+    exec(file.read())
+
+# %%
+from matplotlib import pyplot as plt
+
+from IPython.core.interactiveshell import InteractiveShell
+InteractiveShell.ast_node_interactivity = "all"
+
+# %% [markdown]
+# ### RQ 6.01 What proportion of survey participants offer repairs, servicing and maintenance to heat pump customers
+# - SQ: 42a, 42b
+# - Sample: Standard_output (employment type categorisation)
+
+# %%
+# Print possible answers
+data[col.q42a].value_counts()
+data[col.q42b].value_counts()
+
+
+# %%
+def condition_q42(x):
+    """
+    Function to be applied to dataframe which creates a new column
+    merging responses from sub-types of the same question.
+
+    Args:
+        x: pandas dataframe containing responses from analytical sample.
+    """
+    
+    # Merge responses
+    if (x[col.q42a] == "Yes") or (x[col.q42b] == "Yes"):
+        return "Yes"
+    
+    elif (x[col.q42a] == "No, but I intend to offer it in the next 12 months") or (x[col.q42b] == "No, but we intend to offer it in the next 12 months"):
+        return "No, but we/I intend to offer it in the next 12 months"
+    
+    elif (x[col.q42a] == "No, and I do not intend to start offering this service in the next 12 months") or (x[col.q42b] == "No, and we do not intend to start offering this service in the next 12 months"):
+        return "No, and we/I do not intend to start offering this service in the next 12 months"
+    
+    elif (x[col.q42a] == "Don't know") or (x[col.q42b] == "Don't know"):
+        return "We/I don't know"
+    
+    # Check if all are captured
+    else:
+        return "Flag" 
+
+# Apply function to each row of dataframe
+data["SQ42"] = data.apply(condition_q42,
+                          axis=1
+                          )
+
+# Specifying desired order for outputs
+ans_42 = ["Yes",
+           "No, but we/I intend to offer it in the next 12 months",
+           "No, and we/I do not intend to start offering this service in the next 12 months",
+           "We/I don't know",
+           "Total"
+           ]
+
+# Generate crosstab dataframe
+df_601_crosstab = crosstable("EmploymentType",
```

Comment:
does 601 stand for section 6.01?

Reply:
These functions are introduced from `data_setup.py` but this is not necessarily obvious for someone looking at the script for the first time.  It would probably be good to have a bit more information around what those scripts import.

Reply:
> does 601 stand for section 6.01?

Yes, I named the dataframes and .csv files according to the research question they answer (as listed in the analysis plan which is linked at the start of the notebook).



> These functions are introduced from `data_setup.py` but this is not necessarily obvious for someone looking at the script for the first time. It would probably be good to have a bit more information around what those scripts import.

Thanks for flagging this, I've added some explanatory lines.

```
with open("data_setup.py") as file:
    exec(file.read())
"""
Running "data_setup.py" does the following:
- Imports survey question numbers lookups class as col
- Imports cleaned, anonymised survey analytical sample data as data
- Adds new columns to the data dataframe for each sub-population category,
column names are:
-- EmploymentType
-- CompanySizeOwnerV2
-- CompanySizeEmployeeV2
-- SectorTime
-- NumberInstalls
-- DesiredIncrease
- Defines the following functions to generate cross tables:
-- crosstable(subpop, dataframe, x, ans)
-- location_crosstab(sq_col)
-- explode_select_all(column)
- Defines the following functions to generate figures:
-- stackedbar(df, number, question, section)
-- groupedbar(df, number, question, section)
-- donut(df, number, question, section)
"""
```

```
with open("free_text_recode.py") as file:
    exec(file.read())
"""
Running "free_text_recode.py" re-assigns 'Other' free text responses to Select all that apply type questions
by re-assigning them to a  new or existing answer, or keeping them as 'Other'
"""
```